modules: cmsis, cmsis_6: only add the intended cmsis module

fabiobaltieri opened the pull request:
The current code base is meant to use cmsis for Cortex A and R and cmsis_6 for Cortex M, but the build system is configured to include the path for both when Cortex M is selected. This leaves us exposed to PR using the old headers, that would not get caught in CI but would fail the build on a project using Cortex M that only has the cmsis_6 module.
 
Change the cmsis module setting to only include the module files in the intended case.

Deps
- https://github.com/zephyrproject-rtos/hal_microchip/pull/35
- https://github.com/zephyrproject-rtos/hal_ambiq/pull/56
- https://github.com/zephyrproject-rtos/hal_silabs/pull/122

fabiobaltieri committed:
modules: cmsis, cmsis_6: only add the intended cmsis module

The current code base is meant to use cmsis for Cortex A and R and
cmsis_6 for Cortex M, but the build system is configured to include the
path for both when Cortex M is selected. This leaves us exposed to PR
using the old headers, that would not get caught in CI but would fail
the build on a project using Cortex M that only has the cmsis_6 module.

Change the cmsis module setting to only include the module files in the
intended case.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
ok but it looks like both cmsis-dsp and cmsis-nn have that include line behind a CONFIG entry, so CI should still catch it for tests that do not enable those modules (i.e. most of them, I did run this change before merging the revert and it did indeed fail)

fabiobaltieri commented:
Also, looks like it's not that unused after all :-)

wearyzen commented:
> > Also, looks like it's not that unused after all, even putting those modules aside :-)
> 
> Maybe we introduce a new config USE_CMSIS_5_MODULE which cmsis-nn and cmsis-dsp (or the other current reported modules) enables and is enabled by default for Cortex-A and R, and then the include goes under this config?

Yeah I think this would be good to keep track of who is using the old cmsis as well.

fabiobaltieri committed:
arch: arm: dwt: use the cmsis_6 macro unconditionally

Always use the cmsis_6 version for DWT_LSR_Present_Msk and
DWT_LSR_Access_Msk, the old ones are not going to be available anymore
when Cortex-M is selected..

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Blocked on https://github.com/zephyrproject-rtos/zephyr/issues/93725

fabiobaltieri committed:
manifest: update ambiq and microchip hals to pickup the cmsis_6 fixes

Update various hals to pickup the cmsis_6 related fixes.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

wearyzen commented:
wow that was quick 😃 Looks like you covered everything that Zephyr's ci covers!

fabiobaltieri commented on arch/arm/include/cortex_m/dwt.h:
```diff
@@ -36,15 +36,20 @@ extern "C" {
-#define DWT_LSR_Present_Msk \
-	IF_ENABLED(CONFIG_ZEPHYR_CMSIS_MODULE, (ITM_LSR_Present_Msk)) \
-	IF_DISABLED(CONFIG_ZEPHYR_CMSIS_MODULE, (ITM_LSR_PRESENT_Msk))
+#ifdef CONFIG_CPU_CORTEX_M
```
oh, duh, it's in the name... ok I've redone that commit, getting the modules in may take a while but at least that one is easy

fabiobaltieri commented:
> wow that was quick 😃 Looks like you covered everything that Zephyr's ci covers!

Yeah well, that's what this specific runs cover, there must be more in the hals but I really don't have an effective way of chasing that down so I'm afraid this is the best we can do.

AlessandroLuo commented:
FYI https://github.com/zephyrproject-rtos/hal_ambiq/pull/56 is merged.
